lib/router.js

```diff
@@ -86,138 +140,223 @@ const routes = [
+`key` is not cleared.
+
+  - key. A request key or '*'.
+  - to. An array of components.
+  - from (default: `to`). An array of components.
+  - params. An array of param names.
+*/
+const cacheDataForKey = ({ key, to, params, from = to }) => {
+  for (const toComponent of to) {
+    const { cacheData } = routesByName[toComponent.name].meta;
+    if (cacheData[key] == null) cacheData[key] = {};
+    const paramsByFrom = cacheData[key];
+    for (const fromComponent of from)
+      paramsByFrom[fromComponent.name] = params;
   }
-});
-const state = {
-  navigations: {
-    first: {
-      triggered: false,
-      /* Suppose that during a navigation, a navigation guard uses
-      `next(location)` to redirect the user to a different location, and that
-      redirect navigation is confirmed. According to the standard Vue router
-      terminology, the first navigation is considered aborted, not confirmed,
-      even though it ultimately resulted in a confirmed navigation. For the
-      `confirmed` property, we mean "confirmed" in a looser sense: if a
-      navigation ultimately resulted in a confirmed navigation, the `confirmed`
-      property is true even if a a navigation was aborted along the way. */
-      confirmed: false
-    },
-    last: {
-      triggered: false,
-      // `false` could mean that no navigation has ever been triggered, that the
-      // latest navigation is in progress, or that latest navigation was
-      // ultimately aborted.
-      confirmed: false
-    }
+};
+
+/*
+dataIsCached() returns `true` if the data for `key` should not be cleared when
+the route changes from `from` to `to`. Otherwise it returns `false`.
+
+  - key. A request key or '*'.
+  - to. A Route object.
+  - from. A Route object.
+*/
+const dataIsCached = ({ key, to, from }) => {
+  const { cacheData } = to.matched[to.matched.length - 1].meta;
+  if (cacheData[key] == null) return false;
+  const fromName = from.matched[from.matched.length - 1].name;
+  const params = cacheData[key][fromName];
+  if (params == null) return false;
+  for (const param of params) {
+    if (to.params[param] !== from.params[param]) return false;
   }
+  return true;
 };
 
+// Data that does not change with navigation.
+for (const key of ['session', 'currentUser']) {
+  cacheDataForKey({
+    key,
+    to: namedRoutes.map(route => route.component),
+    params: []
+  });
+}
+
+// Tabs
+cacheDataForKey({
+  key: '*',
+  to: [ProjectOverview, FieldKeyList],
+  params: ['projectId']
+});
+cacheDataForKey({
+  key: '*',
+  to: [FormOverview, FormAttachmentList, FormSubmissionList, FormSettings],
+  params: ['projectId', 'xmlFormId']
+});
+
+cacheDataForKey({
+  key: 'project',
+  to: [
+    // ProjectShow
+    ProjectOverview, FieldKeyList,
+    // FormShow
+    FormOverview, FormAttachmentList, FormSubmissionList, FormSettings
+  ],
+  params: ['projectId']
+});
+
+cacheDataForKey({
+  key: 'project',
+  to: [ProjectOverview],
+  from: [ProjectList],
+  params: []
+});
```

Comment:
> * Is this call of `cacheDataForKey()` a good idea?

This doesn't work here, because while `GET /projects/:id` returns the current user's verbs for the project (when extended metadata is requested),`GET /projects` does not return verbs.

> * Should it work in the reverse direction as well, such that `projects` is cached and is not refreshed when the user navigates back from `ProjectOverview` to `ProjectList`?
* Should we cache the form when moving from `ProjectOverview` to `FormOverview`?

See #204.